cpp/src/io/comp/uncomp.cpp

```diff
@@ -413,167 +390,122 @@ std::vector<char> io_uncompress_single_h2d(const void* src, size_t src_size, int
+               "ZLIB decompression failed");
+  return uncomp_size;
 }
 
 /**
- * @brief ZLIB host decompressor class
+ * @brief GZIP host decompressor (includes header)
  */
-class HostDecompressor_ZLIB : public HostDecompressor {
- public:
-  HostDecompressor_ZLIB(bool gz_hdr_) : gz_hdr(gz_hdr_) {}
-  size_t Decompress(uint8_t* dstBytes,
-                    size_t dstLen,
-                    const uint8_t* srcBytes,
-                    size_t srcLen) override
-  {
-    if (gz_hdr) {
-      gz_archive_s gz;
-      if (!ParseGZArchive(&gz, srcBytes, srcLen)) { return 0; }
-      srcBytes = gz.comp_data;
-      srcLen   = gz.comp_len;
-    }
-    if (0 == cpu_inflate(dstBytes, &dstLen, srcBytes, srcLen)) {
-      return dstLen;
-    } else {
-      return 0;
-    }
-  }
-
- protected:
-  const bool gz_hdr;
-};
+size_t decompress_gzip(host_span<uint8_t const> src, host_span<uint8_t> dst)
+{
+  gz_archive_s gz;
+  auto const parse_succeeded = ParseGZArchive(&gz, src.data(), src.size());
+  CUDF_EXPECTS(parse_succeeded, "Failed to parse GZIP header");
+  return decompress_zlib({gz.comp_data, gz.comp_len}, dst);
+}
 
 /**
- * @brief SNAPPY host decompressor class
+ * @brief SNAPPY host decompressor
  */
-class HostDecompressor_SNAPPY : public HostDecompressor {
- public:
-  HostDecompressor_SNAPPY() {}
-  size_t Decompress(uint8_t* dstBytes,
-                    size_t dstLen,
-                    const uint8_t* srcBytes,
-                    size_t srcLen) override
+size_t decompress_snappy(host_span<uint8_t const> src, host_span<uint8_t> dst)
+{
+  CUDF_EXPECTS(not dst.empty() and src.size() >= 1, "invalid Snappy decompress inputs");
+  uint32_t uncompressed_size, bytes_left, dst_pos;
+  auto cur       = src.begin();
+  auto const end = src.end();
+  // Read uncompressed length (varint)
   {
-    uint32_t uncompressed_size, bytes_left, dst_pos;
-    const uint8_t* cur = srcBytes;
-    const uint8_t* end = srcBytes + srcLen;
-
-    if (!dstBytes || srcLen < 1) { return 0; }
-    // Read uncompressed length (varint)
-    {
-      uint32_t l        = 0, c;
-      uncompressed_size = 0;
-      do {
-        uint32_t lo7;
-        c   = *cur++;
-        lo7 = c & 0x7f;
-        if (l >= 28 && c > 0xf) { return 0; }
-        uncompressed_size |= lo7 << l;
-        l += 7;
-      } while (c > 0x7f && cur < end);
-      if (!uncompressed_size || uncompressed_size > dstLen || cur >= end) {
-        // Destination buffer too small or zero size
-        return 0;
-      }
-    }
-    // Decode lz77
-    dst_pos    = 0;
-    bytes_left = uncompressed_size;
+    uint32_t l        = 0, c;
+    uncompressed_size = 0;
     do {
-      uint32_t blen = *cur++;
-
-      if (blen & 3) {
-        // Copy
-        uint32_t offset;
-        if (blen & 2) {
-          // xxxxxx1x: copy with 6-bit length, 2-byte or 4-byte offset
+      uint32_t lo7;
+      c   = *cur++;
+      lo7 = c & 0x7f;
```

Comment:
Consider this my attempt to find something wrong.  Free free to disregard in the interest of not disturbing the code :) 
```suggestion
      c   = *cur++;
      uint32_t const lo7 = c & 0x7f;
```